packages/beacon-node/src/chain/regen/queued.ts

```diff
@@ -121,6 +121,15 @@ export class QueuedStateRegenerator implements IStateRegenerator {
     return this.checkpointStateCache.updatePreComputedCheckpoint(rootHex, epoch);
   }
 
+  /**
+   * Remove `validators` from all unfinalized cache's epochCtx.UnfinalizedPubkey2Index,
+   * and add them to epochCtx.pubkey2index and epochCtx.index2pubkey
+   */
+  updateUnfinalizedPubkeys(validators: UnfinalizedPubkeyIndexMap): void {
+    this.stateCache.updateUnfinalizedPubkeys(validators);
+    this.stateCache.updateUnfinalizedPubkeys.bind(this.checkpointStateCache)(validators);
```

Comment:
Why is this function called twice in different `this` contexts?

Reply:
> Why is this function called twice in different `this` contexts?

We need to `updateUnfinalizedPubkeys` for `stateCache` and `checkpointStateCache`. Previously `updateUnfinalizedPubkeys` is defined twice in `stateCache` and `checkpointStateCache` with the exact same logic. 

This is an attempt to eliminate code duplication. Maybe there is a better way to do this.

Reply:
this is confusing to me, plus we have different block states/checkpoint states implemented with different data structures inside (new ones are `PersistentCheckpointStateCache`, `FIFOBlockStateCache`)
should be better to add finalized pubkeys once here and do a loop for all state caches to remove unfinalized pubkeys everywhere